### PR TITLE
Tutorial comments and suggestions

### PR DESCRIPTION
**Pull request by sid-kap**
I am not very familiar with pipes---this is only my second time reading through the pipes tutorial, and I've only played around with pipes once or twice.

I expect that my comments might be useful for exactly that reason---that I'm reading the tutorial through a beginner's eyes. Hopefully my concerns might make the tutorial a bit easier for pipes beginners to understand.

Most of my comments fall into the following categories:
- **Motivation:** What is streaming? What is the most obvious way to implement something that is streaming, effectful, and composable? If our solution is different from the obvious implementation, what are the benefits of this approach?
- **Examples:** Sometimes it's hard for a beginner to understand what something does given only a type signature.
- **Explaining types in detail:** When a new type is introduced, I want to see how many type variables the type takes, and what each of these represents (and why they are necessary).
- **Explaining termination:** I think the idea that components can terminate early, and with a return value, should be brought up earlier, and brought up again every time we introduce a new component.

I apologize if some of my suggestions don't belong in the Tutorial. My understanding of the intended audience, and what the reader should expect, might be different from what you intended.

Thanks for all your work on this library! I'm excited to start using Pipes, and I hope the my comments are constructive!

### CODE REVIEW

File: src/Pipes/Tutorial.hs [hunk 1]
```diff
@@ -22,26 +22,34 @@
     transform, and write loop in 'IO', which is streaming and effectful, but is
     not modular or separable.
 
+    At this point, you may be wondering why we can't define a streaming list type as follows:
+@
+newtype Stream m a = Stream (m (Maybe (a, Stream m a)))
+@
+    I am wondering this.
```
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: You can!  This is basically the "`ListT` done right" type.  The `list-t` package provides this type and it's also equivalent to the `Pipes.ListT` type

File: src/Pipes/Tutorial.hs [hunk 2]
```diff
@@ -50,6 +58,7 @@
 
     If you want a Quick Start guide to @pipes@, read the documentation in
     "Pipes.Prelude" from top to bottom.
+    (What, really, is the difference between a Tutorial and a Quick Start guide? Maybe we should be more clear as to which kind of person should read this vs. Pipes.Prelude.)
```
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: For me, a tutorial is something that should leave you with a holistic understanding of how all the pieces fit together.  A Quick Start guide is typically just a bunch of examples that you can cargo cult from with only a superficial understanding of what is going on

File: src/Pipes/Tutorial.hs (hunk 1 above)
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: Hmm. I guess this also ties in to your suggestion later in the tutorial that people live in `ListT` land whenever possible, and then drop down to `Pipes` land when they need lower-level control of termination, return values, etc. 

This would require making the Introduction section longer (which you might not want to do), but I believe it would be nice to describe in a bit more detail up front what our goal is with the Pipes types. 
I haven't thought about this enough, but I think you can achieve all three goals (streaming, effects, composability) using ListT. So, when we tell the reader that we're about to build up all this Pipes machinery, I feel like we should be descriptive about how we went from the three goals (streaming, effects, composability) to Pipes. What additional features does Pipes have beyond the three initial goals?

File: src/Pipes/Tutorial.hs (hunk 2 above)
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: I think when I read this, I missed the sentence that immediately follows:

> This tutorial is more extensive and explains the pipes API in greater detail and illustrates several idioms.

So I guess I'm fine with this as-is.

File: src/Pipes/Tutorial.hs [hunk 3]
```diff
@@ -102,11 +111,13 @@ import Prelude hiding ((.), id)
...
     that you can mix and match diverse stages to produce useful streaming
-    programs.  If you are a library writer, @pipes@ lets you package up
+    programs. [this sentence is a bit wordy---are "diverse" and "useful" necessary here?] If you are a library writer, @pipes@ lets you package up
     streaming components into a reusable interface.  If you are an application
     writer, @pipes@ lets you connect pre-made streaming components with minimal
     effort to produce a highly-efficient program that streams data in constant
-    memory.
+    space.
+
+    What is a stream processing stage? A stream processing stage is a step in the pipeline that either sends data downstream, reads data from upstream, or perform effects in the base monad.
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: Ack, I don't think I conveyed my point well here. Basically what I meant to say was that "stream processing stages" were mentioned in the previous paragraph, but weren't explicitly defined. Combined with the fact that the things you can do with pipes---pushing downstream, reading from upstream, terminating---were not explicitly lined out here or in the introduction, I think it might be confusing. As a reader, I come in with a somewhat vague idea of what "streaming" is (since it's a programming paradigm that isn't used that much yet), so it would be nice to explicitly state our idea of "stream processing" or "streaming components" near the beginning of the tutorial.

File: src/Pipes/Tutorial.hs [hunk 4]
```diff
@@ -115,6 +126,8 @@ import Prelude hiding ((.), id)
 
     * 'await': Receive input data
 
+    Components can also return a value, causing the streaming to terminate immediately and (what really happens to the streamed values that are ignored? what really happens to the return value? when you connect two components, which both return things, do you get both return values?) [When I heard this was a streaming library, I thought it was a type of effectful list, like the `Stream` type I mentioned above. Why are return values necessary?]
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: This ties in with the above note. When I was reading this section and the Producer section, I kind of freaked out when I realized that `Producer` has a `r` type parameter. (The `a` parameter made sense, but the `r` confused me.) It took me a while to grasp the ideas of pipes "returning values" and how this ties into termination. So a lot of my comments in this area are basically suggesting that we talk more, and earlier, about termination and return values.

File: src/Pipes/Tutorial.hs [hunk 5]
```diff
@@ -124,11 +137,9 @@ import Prelude hiding ((.), id)
     * 'Pipe's can both 'yield' and 'await' values and they model stream
       transformations
 
-    * 'Effect's can neither 'yield' nor 'await' and they model non-streaming
-      components
+    * 'Effect's can neither 'yield' nor 'await'---so they're basically equivalent to an action in the base monad
 
-    You can connect these components together in four separate ways which
-    parallel the four above types:
+    In this tutorial, we will explain how you can connect these components together. We will cover four separate ways in which you can connect these components, each of which parallel the four above types:
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: This change is basically trying to make the bulleted list less scary. It means to say, "By the end of this tutorial, you will know about each of these types of composition, so don't worry if you don't understand the contents of the bulleted list right now."

File: src/Pipes/Tutorial.hs [hunk 6]
```diff
@@ -124,11 +137,9 @@ import Prelude hiding ((.), id)
     * 'Pipe's can both 'yield' and 'await' values and they model stream
       transformations
 
-    * 'Effect's can neither 'yield' nor 'await' and they model non-streaming
-      components
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: This relates to my comments above about new readers not knowing the meaning of the term "streaming component". The phrase "non-streaming components" scared me a bit.

File: src/Pipes/Tutorial.hs [hunk 7]
```diff
@@ -124,11 +137,9 @@ import Prelude hiding ((.), id)
...
+    * 'Effect's can neither 'yield' nor 'await'---so they're basically equivalent to an action in the base monad
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: This relates to my comments above about new readers not knowing the meaning of the term "streaming component". The phrase "non-streaming components" scared me a bit.

File: src/Pipes/Tutorial.hs [hunk 8]
```diff
@@ -138,18 +149,35 @@ import Prelude hiding ((.), id)
...
-    monad transformer that extends any base monad with a new 'yield' command.
-    This 'yield' command lets you send output downstream to an anonymous
+    'Producer's are effectful streams of input.  Specifically, a 'Producer a m r' is a
+    monad transformer that extends any base monad 'm' with a new 'yield :: a -> Producer a m r' command.
+    This 'yield' command lets you send output of type 'a' downstream to an anonymous
     handler, decoupling how you generate values from how you consume them.
 
+    At this point, you may be wondering, in a 'Producer a m r', about the
+    purpose of the `r` type. After all, a `Producer` is an "effectful
+    stream of input" of type 'a'. (I'm not sure the answer to this question.)
```
Comment (sid-kap) [inline src/Pipes/Tutorial.hs]: I know now, but I didn't when I was reading this part of the tutorial.

**Comment by sid-kap** [comment]
Bump?

**Comment by Gabriel439** [comment]
Give me some time.  I will get around to this but this requires me to block off a pretty sizable chunk of time to complete

**Comment by sid-kap** [comment]
Sure, no problem.

**Comment by Gabriel439** [comment]
Sorry for the very long delay.  I finally have time to review this (assuming you are still interested)

File: src/Pipes/Tutorial.hs (hunk 1 above)
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: So I recently released a `list-transformer` package which contains basically exactly the type you described and an explanation of the simpler motivation for why to use that type.  It might help to reference that tutorial here

File: src/Pipes/Tutorial.hs [hunk 9]
```diff
@@ -22,26 +22,34 @@
...
+    I am wondering this.
+
     @pipes@ gives you all three features: effectful, streaming, and composable
     programming.  @pipes@ also provides a wide variety of stream programming
     abstractions which are all subsets of a single unified machinery:
 
     * effectful 'Producer's (like generators),
 
-    * effectful 'Consumer's (like iteratees),
+    * effectful 'Consumer's (like iteratees), [maybe we should say what an iteratee is? or is not necessary for this tutorial?]
```
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: Maybe add a link to the Wikipedia article for iteratees here

File: src/Pipes/Tutorial.hs [hunk 10]
```diff
@@ -102,11 +111,13 @@ import Prelude hiding ((.), id)
 {- $introduction
     The @pipes@ library decouples stream processing stages from each other so
     that you can mix and match diverse stages to produce useful streaming
-    programs.  If you are a library writer, @pipes@ lets you package up
+    programs. [this sentence is a bit wordy---are "diverse" and "useful" necessary here?] If you are a library writer, @pipes@ lets you package up
```
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: Yeah, I think "diverse" and "useful" could be omitted

File: src/Pipes/Tutorial.hs (hunk 4 above)
Comment (Gabriel439) [inline src/Pipes/Tutorial.hs]: So to be totally pedantic the `yield`ing a value and `return`ing a value are sort of symmetric with respect to each other.  To see why , take the `Monad` laws for `return` and `(>>=)`:

```haskell
m >>= return = m

return x >>= f = f x

(m >>= f) >>= g = m >>= (\x -> f x >>= g)
```

... and just replace `return` with `yield` and replace `(>>=)` with `for` to get an equivalent set of `Monad` laws:

```haskell
for m yield = m

for (yield x) f = f x

for (for m f) g = for m (\x -> for (f x) g)
```

In fact, these `for`/`yield` laws are basically the `Monad` laws for `ListT`.  If you look at the implementation of the `ListT` type in `pipes` you can see that it's literally just a wrapper that translates `return` to `yield` and `(>>=)` to `for`:

```haskell
instance (Monad m) => Monad (ListT m) where
    return   = pure
    m >>= f  = Select (for (enumerate m) (\a -> enumerate (f a)))
```

So there's really no fundamental difference between `return` and `yield`.  They are just two different ways to emit values, and there are two corresponding ways to capture the values that they emit (i.e. `(>>=)` and `for`, respectively).

**Comment by sid-kap** [comment]
Yes, I'm still interested! I'll look through your comments and respond to them sometime this week.
